### PR TITLE
Support middleware compression with Zlib, LZW, Brotli, and Zstd encoders

### PR DESCRIPTION
**Pull request by AidanKenney**
Issue(s): #197, #198, #199, #200

This pull request provides new Encoders for the Goyave compress middleware using the following algorithms:
- Zlib (deflate)
- LZW
- Brotli
- Zstd

Tests and comments are included

### CODE REVIEW

File: middleware/compress/brotli.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+// NewWriter returns a new `brotli.Writer` using the
+// Compression Quality and LGWin provided in the Brotli encoder
+func (w *Brotli) NewWriter(wr io.Writer) io.WriteCloser {
+	if w.Quality < 0 || w.Quality > 11 {
+		panic(errors.New("Brotli Compression Level must be in range [0, 11]"))
+	}
+	if w.LGWin != 0 {
+		if w.LGWin < 10 || w.LGWin > 24 {
+			panic(errors.New("Brotli LGWin must be either 0 or within range [10, 24]"))
```
Comment (System-Glitch) [inline middleware/compress/brotli.go:33]: Can you add a test for this please?

File: middleware/compress/lzw.go [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+// Lzw encoder for the gzip format using Go's standard `compress/lzw` package.
+//
+// Takes an Order specifying the bit ordering in an LZW data stream level,
+// and number of bits to use for literal codes, litWidth, as parameters.
+// Accepted values are defined by constants
+// in the standard `compress/lzw` package.
+type Lzw struct {
+	Order    lzw.Order
+	LitWidth int
```
Comment (System-Glitch) [inline middleware/compress/lzw.go:18]: Can you please add comments here explaining these two fields and their constraints? You can copy-paste the comments from the `compress/lzw` package. Tell `Order` is optional and its default is `lzw.LSB`.

I'm wondering if we should use a `LitWidth` of 8 as a default if it's not provided, for convenience.

File: middleware/compress/lzw.go [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+}
+
+// Encoding returns "lzw".
+func (w *Lzw) Encoding() string {
+	return "lzw"
```
Comment (System-Glitch) [inline middleware/compress/lzw.go:23]:
```suggestion
	return "compress"
```

File: middleware/compress/zlib.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+// Takes a compression level and "dict" ([]byte) as parameters. Accepted values are defined by constants
+// in the standard `compress/zlib` package.
+type Zlib struct {
+	Dict  []byte
+	Level int
+}
+
+// Encoding returns "zlib".
+func (w *Zlib) Encoding() string {
+	return "zlib"
```
Comment (System-Glitch) [inline middleware/compress/zlib.go:20]:
```suggestion
	return "deflate"
```

File: middleware/compress/zlib.go [hunk 5]
```diff
@@ -0,0 +1,32 @@
+package compress
+
+import (
+	"compress/zlib"
+	"io"
+
+	"goyave.dev/goyave/v5/util/errors"
+)
+
+// Zlib encoder for the gzip format using Go's standard `compress/gzip` package.
```
Comment (System-Glitch) [inline middleware/compress/zlib.go:10]:
```suggestion
// Zlib encoder for the deflate format using Go's standard `compress/zlib` package.
```

File: middleware/compress/lzw.go [hunk 6]
```diff
@@ -0,0 +1,33 @@
+package compress
+
+import (
+	"compress/lzw"
+	"io"
+
+	"goyave.dev/goyave/v5/util/errors"
+)
+
+// Lzw encoder for the gzip format using Go's standard `compress/lzw` package.
```
Comment (System-Glitch) [inline middleware/compress/lzw.go:10]:
```suggestion
// Lzw encoder for the compress format using Go's standard `compress/lzw` package.
```

File: middleware/compress/zlib.go [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+// Takes a compression level and "dict" ([]byte) as parameters. Accepted values are defined by constants
+// in the standard `compress/zlib` package.
+type Zlib struct {
+	Dict  []byte
```
Comment (System-Glitch) [inline middleware/compress/zlib.go:14]: Please add a comment explaining that this field is optional (can be `nil`).

**Review by System-Glitch — CHANGES_REQUESTED** [review]
Four PR in one! Thanks a lot! It just needs a few minor adjustments but otherwise it looks good to me. 👍🏻

File: middleware/compress/zstd_test.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+				},
+			},
+		},
+	}
+
+	request := testutil.NewTestRequest(http.MethodGet, "/zstd", nil)
+	request.Header().Set("Accept-Encoding", "zstd")
+	result := server.TestMiddleware(compressMiddleWare, request, handler)
+
+	reader, _ := zstd.NewReader(result.Body)
```
Comment (ccoVeille) [inline middleware/compress/zstd_test.go:65]:
```suggestion
	reader, err := zstd.NewReader(result.Body)
	require.NoError(t, err)
```

**Comment by AidanKenney** [comment]
Thanks @System-Glitch  @ccoVeille for your reviews! I think I have incorporated everything. 

Tests appears to be failing on /websocket, but I don't think these code changes are related. Please let me know if anything else is needed.

**Comment by System-Glitch** [comment]
> Tests appears to be failing on /websocket
It's the db timeout plugin. It happens sometimes when the runner is a bit too busy. I'm not sure how I could make this test more consistent.

File: middleware/compress/brotli_test.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+		encoder := &Brotli{
+			Quality: 12,
+		}
+		encoder.NewWriter(bytes.NewBuffer([]byte{}))
+	})
+
+	assert.Panics(t, func() {
+		// Invalid LGWin
+		encoder := &Brotli{
+			Quality: 12,
```
Comment (System-Glitch) [inline middleware/compress/brotli_test.go:42]:
```suggestion
			Quality: 11,
```

File: middleware/compress/lzw_test.go [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+func TestLzwEncoder(t *testing.T) {
+	// Default LitWidth of 8 supplied
+	encoder := &LZW{
+		Order: lzw.LSB,
+	}
+
+	assert.Equal(t, "compress", encoder.Encoding())
+
+	buf := bytes.NewBuffer([]byte{})
+	writer := encoder.NewWriter(buf)
```
Comment (System-Glitch) [inline middleware/compress/lzw_test.go:26]:
```suggestion
	writer := encoder.NewWriter(buf)
	assert.Equal(t, 8, encoder.LitWidth)
```

**Review by System-Glitch — APPROVED** [review]
LGTM

**Comment by System-Glitch** [comment]
Merged! Thank you very much. 🎉

**Pull request merged**